src/server.js

```diff
@@ -226,17 +231,23 @@ export class Server extends EventEmitter {
                     if (auth.type !== 'Basic') {
                         throw new RequestError('The "Proxy-Authorization" header must have the "Basic" type.', 400);
                     }
-
                     funcOpts.username = auth.username;
                     funcOpts.password = auth.password;
                 }
-
                 // User function returns a result directly or a promise
                 return this.prepareRequestFunction(funcOpts);
             })
             .then((funcResult) => {
+                this.log(`${result.id} - open connection`);
+
                 // If not authenticated, request client to authenticate
                 if (funcResult && funcResult.requestAuthentication) {
+                    // connection was closed immediately
+                    this.log(`${result.id} - closed connection`);
+                    this.emit('connectionClosed', {
+                        connectionId: result.id,
+                        stats: { srcTxBytes: 0, srcRxBytes: 0 },
+                    });
                     throw new RequestError('Proxy credentials required.', 407);
```

Comment:
Actually I thought that 'connectionClosed' shouldn't be emitted more than once as it will be confusing for callers. Here, for example, the event will be emitted once and then RequestError will cause failRequest() to emit the event again.

Reply:
connectionClosed here is emitted because no handler was actually created (it fails in constructor of handler) but it is already reported to the main proxy component in prepareRequestFunction. Basically this is just a cleanup call. If this wasn't here than any calls that correctly assign auth options but incorrectly set upstreamProxy would fail and the connection will be never marked as closed.

Reply:
Yeah, no handler is created but `this.failRequest(request, err, handlerOptions);` will be called anyway, in both `onRequest()` and `onConnect()`, and it emits the event again

Reply:
Unless I'm mistaken for each connection either onRequest or onCalled is called, not both. And for each function it can call failRequest 3 times:

1) When prepareRequestFunction throws error, then handlerOptions is empty and failRequest does not emit "connectionClosed"
2) When handler is constructed, which will call failRequest with handlerOptions, so it will emit "connectionClosed" 
3) When it fails during runHandler, which results in the same as during constructor. 

Only if both constructor and runHandler are processed, are events for 'close' and 'end'  processed which in turn can call "connectionClosed", so there is no way that both failRequest and Handler.emitHandlerClosed are called.

I added an unit test which checks it this for all available tests and it looks like it works.

Reply:
Oh I see, sorry, didn't notice this before.